crates/subspace-farmer/src/plot.rs

```diff
@@ -344,51 +344,65 @@ impl Plot {
-
-        self.inner
-            .requests_sender
-            .send(RequestWithPriority {
-                request: Request::ReadPieceIndexes {
-                    from_index_hash,
-                    count,
-                    result_sender,
-                },
-                priority: RequestPriority::Low,
-            })
-            .map_err(|error| {
-                io::Error::other(format!(
-                    "Failed sending read piece indexes request: {error}"
-                ))
-            })?;
-
-        result_receiver.recv().map_err(|error| {
-            io::Error::other(format!(
-                "Read piece indexes result sender was dropped: {error}",
-            ))
-        })?
-    }
+    const PIECES_PER_REQUEST: u64 = 1000;
 
     // TODO: Return (Vec<PieceIndex>, FlatPieces) instead
     /// Returns pieces and their indexes starting from supplied piece index hash (`from`)
     pub(crate) fn get_sequential_pieces(
         &self,
-        from: PieceIndexHash,
+        mut from_index_hash: PieceIndexHash,
         count: u64,
     ) -> io::Result<Vec<(PieceIndex, Piece)>> {
-        self.read_sequential_piece_indexes(from, count)?
-            .into_iter()
-            .map(|index| {
-                self.read_piece(PieceIndexHash::from_index(index))
-                    .map(|piece| (index, piece))
-            })
-            .collect()
+        let (result_sender, result_receiver) = mpsc::channel();
+        let receive_pieces = |from_index_hash, count| {
+            self.inner
+                .requests_sender
+                .send(RequestWithPriority {
+                    request: Request::ReadSequentialPieces {
+                        from_index_hash,
+                        count,
+                        result_sender: result_sender.clone(),
+                    },
+                    priority: RequestPriority::Low,
+                })
+                .map_err(|error| {
+                    io::Error::other(format!(
+                        "Failed sending read piece indexes request: {error}"
+                    ))
+                })?;
+
+            result_receiver.recv().map_err(|error| {
+                io::Error::other(format!(
+                    "Read piece indexes result sender was dropped: {error}",
+                ))
+            })?
+        };
+
+        let mut pieces = Vec::with_capacity(count as usize);
+        for count in std::iter::repeat(Self::PIECES_PER_REQUEST)
+            .take((count / Self::PIECES_PER_REQUEST) as usize)
+        {
+            let mut new_pieces = receive_pieces(from_index_hash, count)?;
+            if count > new_pieces.len() as u64 {
+                pieces.extend(new_pieces);
+                return Ok(pieces);
+            }
+
+            from_index_hash = new_pieces
+                .pop()
+                .map(|(idx, _)| PieceIndexHash::from_index(idx))
+                .expect("Vector is not empty");
```

Comment:
You're removing an element here that will not be inserted into `pieces`, did you intend to do that? Seems like a bug to me.

Reply:
Yep, it is intended in order to not return the same piece twice.

Reply:
Why not adding `+1` to the next request and avoid reading the same piece twice in the first place? Strange and error-prone approach :confused:

Reply:
Oh, you right, haha :)
Let me fix that